### ISSUE TITLE
Do not maintain a connection to the camera if not required

### ISSUE THREAD

**Issue by felipecrs** [blakeblackshear/frigate#4325]
**Describe what you are trying to accomplish and why in non technical terms**

I have a battery-powered camera in which I want to add to Frigate, but I don't want Frigate to record or detect objects on it all the time. I will enable Frigate's recording dynamically through external events, like by an external motion sensor in Home Assistant.

When there's no one connected in the camera, the camera enters in sleep mode to save battery.

**Describe the solution you'd like**
Ideally Frigate should NOT maintain a connection with the camera for a configuration like below:

```yaml
video_porteiro:
    ffmpeg:
      inputs:
        - path: rtsp://192.168.1.10:8554/camera.video_porteiro
          roles:
            - detect
            - record
            - rtmp
      output_args:
        record: -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -c:v copy -c:a aac
        rtmp: -c:v copy -c:a aac -f flv
    record:
      enabled: false
    rtmp:
      enabled: true
    detect:
      enabled: false
      width: 640
      height: 360
    live:
      height: 360
```

I have `rtmp.enabled: true`, but I have no client streaming from it (and `preload` is not activated in Home Assistant configuration). Frigate should understand that a connection with the camera is not required in this situation.

**Describe alternatives you've considered**
I can't think of any.

**Additional context**
I'm using go2rtc to capture my Tuya battery-powered video doorbell through this:

- https://github.com/felipecrs/hass-expose-camera-stream-source#importing-home-assistant-cameras-to-go2rtc-andor-frigate

And I'm reusing the go2rtc stream in Frigate. In the go2rtc info panel I can see that it's Frigate  who is connected to the stream.

**Comment by NickM-27** [comment]
This seems like a somewhat duplicate or at least would be covered by https://github.com/blakeblackshear/frigate/issues/1911

**Comment by felipecrs** [comment]
Are you sure? Sorry I didn't read the whole issue, but I don't think I want to "deactivate" the camera dynamically. I want the camera to be available for streaming all the time, but I don't want it to be "preloaded". When there's no one streaming from it, Frigate should stop the connection with the camera.

**Comment by felipecrs** [comment]
Just to add:

![image](https://user-images.githubusercontent.com/29582865/200661668-433ecfd8-6f49-4ad9-8e01-78ecb23afc86.png)

I have nothing enabled for the camera ATM.

**Comment by NickM-27** [comment]
> Are you sure?

That feature request is after the exact same thing: Not having an ffmpeg process /  stream open on a camera when it is not being used for anything. 

The only difference that I see between the two requests is that you're suggesting an implicit "toggle" that is enabled when one or more processes that need the stream are enabled meanwhile the other request is asking for an explicit toggle to tell frigate when this is the case. If either is implemented the other becomes redundant.

**Comment by felipecrs** [comment]
Well, the manual toggle situation in which I would need to disable the camera after using it and enabling it before using it doesn't fulfill my use case, but yes, we can simply amend #1911 mentioning such implicit toggle:

- Camera auto-deactivates when detect, motion, and recording is disabled and there's no client connected in the restream interface.
- Camera auto-activates when detect, motion, or recording is enabled, or a client connects in the restream interface.

**Comment by felipecrs** [comment]
I say auto-detect because there might be some other situations in which user may want to manually disable the camera, e.g. disable the bedroom camera entirely when a person is inside.

**Comment by NickM-27** [comment]
> Well, the manual toggle situation in which I would need to disable the camera after using it and enabling it before using it doesn't fulfill my use case

Not necessarily, for example if a camera is currently disabled and you want recording then turning recording on could also just enable the camera if it is not already. Similar for the other way around. 

Either way these are implementation details 👍

**Comment by felipecrs** [comment]
I understand that down to the implementation these are only details, but from the user perspective, I think it changes everything. I think the ideal UX would be something like:

1. User has a toggle to enable/disable camera, so that he can build automations like only enable bedroom camera when everyone is away from home.
2. When disabled, camera changes its state to `off`, `disabled`.
3. Camera auto-enters in `idle` state when there's no consumer (from user perspective this has nothing to do with enabling/disabling the camera)
5. Camera auto-enters in `recording`, `streaming`, or other states when recording is enabled, or there's someone consuming its stream, respectively.
6. If camera is manually disabled, it will never auto-enter in `recording`, `streaming` state.

TL;DR:

#1911 suggests `on`, `off` states and switches to camera
#4325 (this) suggests `idle` state when there's no consumer

**Comment by NickM-27** [comment]
I don't see why these can't co-exist. The way I see it there's an explicit toggle that is essentially a quick link to disabling recording, detect, and motion. Enabling it will enable the services that were enabled previously. 

Similarly if you disable recording, detect, motion then the camera is functionally disabled. 

> If camera is manually disabled, it will never auto-enter in recording, streaming state.

The streaming part is not possible currently given that we are switching to go2rtc for restreaming, it does not provide a way to disable the stream.

**Comment by NickM-27** [comment]
You said "if a camera is disabled it will not auto enter streaming state" and I'm saying this isn't possible, there's no way to force go2rtc to not start streaming if a consumer requests

**Comment by felipecrs** [comment]
Well... there is. Frigate can supply a `NO VIDEO` or `Camera is disabled` picture as its input.

**Comment by felipecrs** [comment]
Well, Frigate can also delete the stream from go2rtc as well.

**Comment by NickM-27** [comment]
> Well... there is. Frigate can supply a `NO VIDEO` or `Camera is disabled` picture as its input.

Frigate has no control or knowledge of the consumer request for restream from go2rtc. 

> Well, Frigate can also delete the stream from go2rtc as well (I suppose there's a API for it, since there's an API to add).

Yeah that's true

**Comment by NickM-27** [comment]
Given that I think there's enough distinction to warrant separate requests though they may be addressed in the same solution.

**Comment by chatziko** [comment]
With the go2rtc integration and the rtmp role being deprecated, this should be easier to implement. No need to monitor clients viewing the stream, this is already done by go2rtc.

Now frigate is only involved in recording/detection and these functionalities are only off when the corresponding switch is off. So the goal would be to close the detector ffmpeg process when detection is disabled, and similary for record.

**Comment by felipecrs** [comment]
Just a question, is the fix for this problem aimed to Frigate 0.13?

**Comment by NickM-27** [comment]
> Just a question, is Frigate 0.13 aiming to fix this problem?

Not for 0.13

**Comment by felipecrs** [comment]
Ok, thank you.

**Comment by felipecrs** [comment]
Just to explain another use case for this feature:

@pergolafabio installed Frigate not because of object detection or recording, but to be able to use 2-way audio with Frigate Card through go2rtc with his doorbell. But he realized that Frigate maintains a  ffmpeg process with the camera even though there is nothing consuming from it, and it's eating up too much CPU.

Another solution for this problem would be

- https://github.com/dermotduffy/frigate-hass-card/issues/1206

But I think it's a good thing for Frigate to be as lightweight as possible anyway. Having access to the Frigate integration for example is very handy.

**Comment by NickM-27** [comment]
It is certainly a valid usecase, it is just that this idea is quite complicated in practice, so it needs to be done carefully and thoughtfully. See https://github.com/blakeblackshear/frigate/pull/6744#issuecomment-1634654897 for more context

**Comment by Sammy1Am** [comment]
I know this is an old issue, but I've also come across a similar use-case where this would be handy so I thought I'd (hopefully not annoyingly) bump this issue.

Most of our cameras fit Frigate's typical usecase for detection, but there a few that we don't really need/want any recording/detection on, but would still like to be able to check in on on-demand.  It's possible to just connect to these via Frigate's go2rtc instance using VLC, etc., but (especially in 0.14) Frigate's UI is really awesome and it would be great to be able to use that for consistency/convenience.

Even more so once user roles get implemented, it would be nice to use Frigate as a one-stop-shop for viewing cameras (in addition to all its detection/recording features).

**Comment by NickM-27** [comment]
Now that #1911 is fully implemented, I am going to close this as well. Disabling all features on a camera is effectively the same thing as leaving those features enabled and just turning off the camera, but with fewer steps. It is the correct way to signal to Frigate that nothing is needed from it right now. This also simplifies things like that Frigate does not need to do book-keeping of go2rtc streams which would add a lot of complexity.

**Comment by felipecrs** [comment]
I don't think you should close this, and here's why:

I never said I wanted to disable the camera (which was possible since ever through `config.yaml` anyway).

This feature is about not maintaining connections to the camera when not necessary.

I.e. if `detect`, `motion`, `record`, `birdseye`, and `audio` are all disabled, Frigate was not supposed to maintain a connection with the camera.

But the camera would still be enabled, so that I could live view it from the Frigate UI.

**Comment by Sammy1Am** [comment]
I tried reading through #1911 and #16894 , but I need some help connecting the dots to this issue if you don't mind (if there's some documentation somewhere for 0.16 on the new changes, feel free to just point me there instead).

#16894 says that when a camera is "disabled" (which is different from `enabled: false` I think?) that the ffmpeg process will be stopped, which *is* more or less what this issue was asking about.  But when I open up the UI, would the camera still be visible there?  My understanding is that depending on the configuration, ffmpeg is not really part of the camera -> UI pipeline, so this seems plausible, but I just want to check.

**Comment by hawkeye217** [comment]
> https://github.com/blakeblackshear/frigate/pull/16894 says that when a camera is "disabled" (which is different from enabled: false I think?) that the ffmpeg process will be stopped... But when I open up the UI, would the camera still be visible there? 

No, when a camera is disabled, Frigate stops any ffmpeg processes that are connected to it. If your stream is served by go2rtc and there are no other clients connected to it, go2rtc will stop consuming the feed from your camera and use little to no resources.

**Comment by NickM-27** [comment]
> But the camera would still be enabled, so that I could live view it from the Frigate UI.

Using the Frigate UI requires an active connection to the camera, whether that is for decoding the stream to use for the still images on the camera grid or when actively live viewing. 

That would essentially mean that this request asks that when `detect` `motion` `record` `birdseye` and `audio` are disabled for a camera then it should half break the UI for the live dashboard, which I don't think is a good user experience and would add additional complexity to handling different camera states. 

and realistically we don't need all of this extra complex logic anyway, simply having a button in the live UI to force a live view despite the camera being enabled would solve this concern as well.

**Comment by Sammy1Am** [comment]
> The UI will also show a "Camera is disabled" message over a blank box. If your stream is served by go2rtc and there are no other clients connected to it, go2rtc will also stop consuming the feed from your camera and use little to no resources.

Am I incorrect in my understanding that the UI uses go2rtc directly (e.g. via WebRTC) to display the camera feeds?  If so, why not allow the UI to do this instead of showing "Camera is disabled"?

> simply having a button in the live single camera UI to force a live view despite the camera being disabled would solve this concern as well.

Rather than require pressing a button every single time I want to view the camera feed, it would be convenient if when I loaded up the UI to view the camera it automatically connected and displayed the feed the way it does currently.

**Comment by hawkeye217** [comment]
> Am I incorrect in my understanding that the UI uses go2rtc directly (e.g. via WebRTC) to display the camera feeds? If so, why not allow the UI to do this instead of showing "Camera is disabled"?

Yes, that is incorrect. By default without go2rtc, Frigate uses images from the `detect` stream along with the `detect` stream itself for live view on the dashboard, and exclusively the `detect` feed on the single camera live view. If go2rtc is configured, the `detect` stream is still used on the dashboard. On the single camera live view, the `detect` stream is the fallback if go2rtc's MSE is non-functional and if a user does not have WebRTC configured.

So to load the live feed "the way it does currently", the `detect` stream (where Frigate decodes using ffmpeg) is required.

**Comment by Sammy1Am** [comment]
Ah, okay, thank you for the clarification, some of this makes more sense now.

> By default without go2rtc

Honestly it hadn't even occurred to me to *not* use go2rtc for the cameras, but I do see now that it's optional.

> Frigate uses images from the `detect` stream along with the `detect` stream itself for live view on the dashboard

This was surprising to me, and (unrelated to this issue) maybe could be documented better.  I assumed the `detect` role would potentially be downsized to optimize resources for detection, but `record` could be e.g. higher resolution for quality viewing.  However, knowing that all the live views are dependent on `detect`, I now understand the connection between this issue and the others.

> we don't need all of this extra complex logic anyway, simply having a button in the live single camera UI to force a live view despite the camera being disabled would solve this concern as well.

It seems like if such a button can exist, it would also be technically possible for the UI to automatically restart whatever process is needed to show a live view of the camera when the page was loaded (and stop the process afterward if the camera is still disabled). I agree no complex logic is required though, it's enough to just disable the camera to say "don't do any processing of this".

-----

My dream goal with this issue was basically being able to view camera feeds on the Frigate dashboard even if those cameras weren't being managed / processed otherwise by Frigate because a) it would be convenient to have all my camera feeds in one place, and b) I think the Frigate UI is much nicer than e.g. a Home Assistant dashboard (which can technically fulfill my goal).  However if the Frigate dashboard is specifically designed to show `detect` streams as a way to monitor just the object/motion detection processes, I understand how this is out of scope.

Thank you again for taking the time to clarify things.

**Comment by hawkeye217** [comment]
> This was surprising to me, and (unrelated to this issue) maybe could be documented better.

Have a look at the FAQ question 5 in the dev Live view docs where this behavior is documented: https://deploy-preview-16390--frigate-docs.netlify.app/configuration/live#live-view-faq

**Comment by felipecrs** [comment]
> Using the Frigate UI requires an active connection to the camera, whether that is for decoding the stream to use for the still images on the camera grid or when actively live viewing.

Yes, but that would be on-demand. When the Frigate UI is open and the camera is being viewed, then Frigate should connect to the camera. UI can handle "loading" while the connection is being established.

Then, after all clients disconnect, the connection can be broken and then there should be no connection to the camera (and it can be established again if the Frigate UI is reopened in the future).

---

I understand if you close this issue with _Sorry, this is too complex, and because it's too complex we don't want to spend time on it_.

I just don't think closing in favor/because of #1911 is correct.

**Comment by NickM-27** [comment]
> My dream goal with this issue was basically being able to view camera feeds on the Frigate dashboard even if those cameras weren't being managed / processed otherwise by Frigate

To be clear you don't have to have frigate processing enabled like object detection and everything. You can disable everything so that frigate only does decoding (lightweight when you have an iGPU). It maintains an active connection to the camera, but realistically I think that's only a problem when using a battery camera.

**Comment by felipecrs** [comment]
> It maintains an active connection to the camera, but realistically I think that's only a problem when using a battery camera.

That's exactly why I opened this issue. It's in the first line. :)

**Comment by NickM-27** [comment]
Right, and frigate is not designed to work with battery cameras, so that is outside of the scope.
